AK/Error.h

```diff
@@ -149,62 +144,142 @@ class [[nodiscard]] ErrorOr {
 
     template<typename U>
-    ALWAYS_INLINE ErrorOr(U&& value)
-    requires(
-        requires { T(declval<U>()); } || requires { ErrorType(declval<RemoveCVReference<U>>()); })
-        : m_value_or_error(forward<U>(value))
+    constexpr ALWAYS_INLINE ErrorOr(U&& value)
+    requires(requires { T(declval<U>()); } && !IsSame<U, ErrorType>)
+        : m_value(forward<U>(value))
+        , m_is_error(false)
+    {
+    }
+    template<typename U>
+    constexpr ALWAYS_INLINE ErrorOr(U&& error)
+    requires(requires { ErrorType(declval<RemoveCVReference<U>>()); } && !IsSame<U, T>)
+        : m_error(forward<U>(error))
+        , m_is_error(true)
     {
     }
 
 #ifdef AK_OS_SERENITY
-    ErrorOr(ErrnoCode code)
-        : m_value_or_error(Error::from_errno(code))
+    constexpr ALWAYS_INLINE ErrorOr(ErrnoCode code)
+        : m_error(Error::from_errno(code))
+        , m_is_error(true)
     {
     }
 #endif
 
-    T& value()
+    constexpr ALWAYS_INLINE T& value()
+    {
+        VERIFY(!is_error());
+        return m_value;
+    }
+    constexpr ALWAYS_INLINE T const& value() const
     {
-        return m_value_or_error.template get<T>();
+        VERIFY(!is_error());
+        return m_value;
     }
-    T const& value() const { return m_value_or_error.template get<T>(); }
 
-    ErrorType& error() { return m_value_or_error.template get<ErrorType>(); }
-    ErrorType const& error() const { return m_value_or_error.template get<ErrorType>(); }
+    constexpr ALWAYS_INLINE ErrorType& error()
+    {
+        VERIFY(is_error());
+        return m_error;
+    }
+    constexpr ALWAYS_INLINE ErrorType const& error() const
+    {
+        VERIFY(is_error());
+        return m_error;
+    }
 
-    bool is_error() const { return m_value_or_error.template has<ErrorType>(); }
+    constexpr ALWAYS_INLINE bool is_error() const { return m_is_error; }
 
-    T release_value() { return move(value()); }
-    ErrorType release_error() { return move(error()); }
+    constexpr ALWAYS_INLINE T release_value() { return move(value()); }
+    constexpr ALWAYS_INLINE ErrorType release_error() { return move(error()); }
 
-    T release_value_but_fixme_should_propagate_errors()
+    constexpr ALWAYS_INLINE T release_value_but_fixme_should_propagate_errors()
     {
         VERIFY(!is_error());
         return release_value();
     }
 
+    ~ErrorOr()
+    requires(!Detail::IsDestructible<T> || !Detail::IsDestructible<ErrorType>)
+    = delete;
+
+    constexpr ALWAYS_INLINE ~ErrorOr()
+    requires(IsTriviallyDestructible<T> && IsTriviallyDestructible<ErrorType>)
+    = default;
+    constexpr ALWAYS_INLINE ~ErrorOr()
+    {
+        if (m_is_error)
+            m_error.~ErrorType();
+        else
+            m_value.~T();
+    }
+
 private:
-    Variant<T, ErrorType> m_value_or_error;
+#ifdef AK_COMPILER_CLANG
+    // FIXME: This is a workaround for the fact that the Variant based approach allowed this, somehow
+    //        And at least one users triggers this
```

Comment:
What, how was it even working?

Where's that usage?

Reply:
In the USB stack
Variant allows it as it uses a `u8 m_storage[max(sizeof(Ts)...)]` which evaluates the size of trailing variable arrays as 0,
The user in question I found is an enum representing a buffer we might get returned by HW, it has a member emulating a max sized buffer, so it should be safe (in USB UAS)

Reply:
Ok now it get's funny:
GCC doesnt allow it at all?